internal/driver/controllerserver_helper.go

```diff
@@ -0,0 +1,328 @@
+	if err != nil {
+		return 0, errInternal("list instance disks: %v", err)
+	}
+
+	// Convert the reported memory from MB to bytes
+	memBytes := uint(instance.Specs.Memory) << 20
+	return maxVolumeAttachments(memBytes) - len(disks), nil
+}
+
+// attemptGetContentSourceVolume retrieves information about the Linode volume to clone from.
+// It returns a LinodeVolumeKey if a valid source volume is found, or an error if the source is invalid.
+func (cs *ControllerServer) attemptGetContentSourceVolume(ctx context.Context, contentSource *csi.VolumeContentSource) (*linodevolumes.LinodeVolumeKey, error) {
+	log := logger.GetLogger(ctx)
+	log.V(4).Info("Attempting to get content source volume")
+
+	if contentSource == nil {
+		return nil, nil // Return nil if no content source is provided
+	}
+
+	// Check if the content source type is a volume
+	if _, ok := contentSource.GetType().(*csi.VolumeContentSource_Volume); !ok {
+		return nil, errUnsupportedVolumeContentSource
+	}
+
+	sourceVolume := contentSource.GetVolume()
+	if sourceVolume == nil {
+		return nil, errNoSourceVolume // Return error if no source volume is specified
+	}
+
+	// Parse the volume ID from the content source
+	volumeInfo, err := linodevolumes.ParseLinodeVolumeKey(sourceVolume.GetVolumeId())
+	if err != nil {
+		return nil, errInternal("parse volume info from content source: %v", err)
+	}
+
+	// Retrieve the volume data using the parsed volume ID
+	volumeData, err := cs.client.GetVolume(ctx, volumeInfo.VolumeID)
+	if err != nil {
+		return nil, errInternal("get volume %d: %v", volumeInfo.VolumeID, err)
+	}
+
+	// Check if the volume's region matches the server's metadata region
+	if volumeData.Region != cs.metadata.Region {
+		return nil, errRegionMismatch(volumeData.Region, cs.metadata.Region)
+	}
+
+	return volumeInfo, nil // Return the parsed volume information
+}
+
+// attemptCreateLinodeVolume creates a Linode volume while ensuring idempotency.
+// It checks for existing volumes with the same label and either returns the existing
+// volume or creates a new one, optionally cloning from a source volume.
+func (cs *ControllerServer) attemptCreateLinodeVolume(ctx context.Context, label string, sizeGB int, tags string, sourceVolume *linodevolumes.LinodeVolumeKey) (*linodego.Volume, error) {
+	log := logger.GetLogger(ctx)
+	log.V(4).Info("Attempting to create Linode volume", "label", label, "sizeGB", sizeGB, "tags", tags)
+
+	// List existing volumes with the specified label
+	jsonFilter, err := json.Marshal(map[string]string{"label": label})
+	if err != nil {
+		return nil, errInternal("marshal json filter: %v", err)
+	}
+
+	volumes, err := cs.client.ListVolumes(ctx, linodego.NewListOptions(0, string(jsonFilter)))
+	if err != nil {
+		return nil, errInternal("list volumes: %v", err)
+	}
+
+	// Raise an error if more than one volume with the same label exists
+	if len(volumes) > 1 {
+		return nil, errAlreadyExists("volume %q already exists", label)
```

Comment:
this error message doesnt match the comment above. if multiple volumes are found, the error message should reflect that

Reply:
ah good catch!